cf_xarray/accessor.py

```diff
@@ -2744,22 +2745,128 @@ def __getitem__(self, key: Hashable | Iterable[Hashable]) -> DataArray:
 
         return _getitem(self, key)
 
+    def __eq__(self, other: object):
+        """
+        Compare flag values against `other`.
+
+        `other` must be in the 'flag_meanings' attribute.
+        """
+        if other not in self.flags.data_vars:
+            raise ValueError(
+                f"Flag meaning {other!r} not in known meanings "
+                f"{list(self.flags.data_vars)!r}"
+            )
+        return self.flags[other].rename(self._obj.name)
+
+    def __ne__(self, other: object):
+        """
+        Compare flag values against `other`.
+
+        `other` must be in the 'flag_meanings' attribute.
+        """
+        if other not in self.flags.data_vars:
+            raise ValueError(
+                f"Flag meaning {other!r} not in known meanings "
+                f"{list(self.flags.data_vars)!r}"
+            )
+        return ~self.flags[other].rename(self._obj.name)
+
+    @property
+    def flags(self) -> Dataset:
+        """
+        Dataset containing boolean masks of available flags.
+        """
+        return self._extract_flags()
+
+    def _extract_flags(self, flags: Sequence[Hashable] | None = None) -> Dataset:
```

Comment:
There is an alternative implementation here: https://nbviewer.org/gist/scottyhq/efd583d66999ce8f6e8bcefa81545b8d

```python
def extract_masks(da):
    ''' given a da.fmask return all mask arrays as an xarray dataset'''
    da = da.astype('uint8')
    dsM = da.to_dataset()
    
    explode = (da.shape[0], da.shape[1], 8)
    masks = np.unpackbits(da, axis=1).reshape(explode)
    
    aerosols = masks[:,:,0:2].sum(axis=-1)
    high_aerosol = (aerosols == 3)
    moderate_aerosol = (aerosols == 2)
    low_aerosol = (aerosols == 1)

    # Now to get a mask array it looks like this:
    masks = masks.astype('bool')
    water = masks[:,:,2]
    snow = masks[:,:,3]
    cloud_shadow = masks[:,:,4]
    cloud_adjacent = masks[:,:,5]
    cloud = masks[:,:,6]
    
    dsM['water'] = (('y','x'), water)
    dsM['snow'] = (('y','x'), snow)
    dsM['cloud_shadow'] = (('y','x'), cloud_shadow)
    dsM['cloud_adjacent'] = (('y','x'), cloud_adjacent)
    dsM['cloud'] = (('y','x'), cloud)
    dsM['high_aerosol'] = (('y','x'), high_aerosol)
    dsM['mod_aerosol'] = (('y','x'), moderate_aerosol)
    dsM['low_aerosol'] = (('y','x'), low_aerosol)
    dsM = dsM.drop_vars('Fmask')

    return dsM
```

for data

```
(LSB to MSB)

0: cirrus
1: cloud
2: cloud/shadow adjacent
3: cloud shadow
4: snow/isce
5: water
6-7: aerosol level
    - 11=high aerosol (3)
    - 10 moderate (2)
    - 01 low (1)
    - 00 climatrology aerosol (0)
```

Reply:
I discarded np.unpackbits early when prototyping since it only works for 8 bits at most.
I am not sure if it is substantially faster than our current approach of applying a bitwise AND. The bitwise operation also directly allows for easy comparison when combining 'flag_values' and 'flag_masks', in contrast to here where additional computation is needed for aerosols (and where I think there is an error, low and moderate are not differentiated).
Completely unpacking bits is a nice approach, it's very clear to read and understand, but has some pitfalls I think.

Reply:
Oh hmm.. that's a good point.

Can you link to some real dataset you use with this feature?